lib/Dialect/CKKS/IR/CKKSDialect.cpp

```diff
@@ -0,0 +1,148 @@
+#include "lib/Dialect/CKKS/IR/CKKSDialect.h"
+
+#include <optional>
+
+#include "lib/Dialect/CKKS/IR/CKKSOps.h"
+#include "lib/Dialect/LWE/IR/LWEAttributes.h"
+#include "lib/Dialect/LWE/IR/LWETypes.h"
+#include "llvm/include/llvm/ADT/TypeSwitch.h"            // from @llvm-project
+#include "llvm/include/llvm/Support/ErrorHandling.h"     // from @llvm-project
+#include "mlir/include/mlir/IR/Builders.h"               // from @llvm-project
+#include "mlir/include/mlir/IR/DialectImplementation.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/Location.h"               // from @llvm-project
+#include "mlir/include/mlir/IR/MLIRContext.h"            // from @llvm-project
+#include "mlir/include/mlir/Support/LLVM.h"              // from @llvm-project
+
+// Generated definitions
+#include "lib/Dialect/CKKS/IR/CKKSDialect.cpp.inc"
+#include "mlir/include/mlir/Support/LogicalResult.h"  // from @llvm-project
+#define GET_OP_CLASSES
+#include "lib/Dialect/CKKS/IR/CKKSOps.cpp.inc"
+
+namespace mlir {
+namespace heir {
+namespace ckks {
+
+//===----------------------------------------------------------------------===//
+// CKKS dialect.
+//===----------------------------------------------------------------------===//
+
+// Dialect construction: there is one instance per context and it registers its
+// operations, types, and interfaces here.
+void CKKSDialect::initialize() {
+  addOperations<
+#define GET_OP_LIST
+#include "lib/Dialect/CKKS/IR/CKKSOps.cpp.inc"
+      >();
+}
+
+LogicalResult MulOp::verify() {
+  auto x = getLhs().getType();
+  auto y = getRhs().getType();
+  if (x.getRlweParams().getDimension() != y.getRlweParams().getDimension()) {
+    return emitOpError() << "input dimensions do not match";
+  }
+  auto out = getOutput().getType();
+  if (out.getRlweParams().getDimension() !=
+      y.getRlweParams().getDimension() + x.getRlweParams().getDimension() - 1) {
+    return emitOpError() << "output.dim == x.dim + y.dim - 1 does not hold";
+  }
+  return success();
+}
+
+LogicalResult RotateOp::verify() {
+  auto x = getInput().getType();
+  if (x.getRlweParams().getDimension() != 2) {
+    return emitOpError() << "x.dim == 2 does not hold";
+  }
+  auto out = getOutput().getType();
+  if (out.getRlweParams().getDimension() != 2) {
+    return emitOpError() << "output.dim == 2 does not hold";
+  }
+  return success();
+}
+
+LogicalResult RelinearizeOp::verify() {
+  auto x = getInput().getType();
+  auto out = getOutput().getType();
+  if (x.getRlweParams().getDimension() != getFromBasis().size()) {
+    return emitOpError() << "input dimension does not match from_basis";
+  }
+  if (out.getRlweParams().getDimension() != getToBasis().size()) {
+    return emitOpError() << "output dimension does not match to_basis";
+  }
+  return success();
```

Comment:
We could/should move stuff like this (which is also present in BGV/BFV) to some kind of common utilities file